{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: pymongo in c:\\users\\hp\\anaconda3\\lib\\site-packages (4.10.1)\n",
      "Requirement already satisfied: dnspython<3.0.0,>=1.16.0 in c:\\users\\hp\\anaconda3\\lib\\site-packages (from pymongo) (2.7.0)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\utils\\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedWriter name=5>\n",
      "  return process_handler(cmd, _system_body)\n",
      "ResourceWarning: Enable tracemalloc to get the object allocation traceback\n",
      "c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\utils\\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedReader name=6>\n",
      "  return process_handler(cmd, _system_body)\n",
      "ResourceWarning: Enable tracemalloc to get the object allocation traceback\n",
      "c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\utils\\_process_win32.py:124: ResourceWarning: unclosed file <_io.BufferedReader name=7>\n",
      "  return process_handler(cmd, _system_body)\n",
      "ResourceWarning: Enable tracemalloc to get the object allocation traceback\n"
     ]
    }
   ],
   "source": [
    "!pip install pymongo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hp\\AppData\\Local\\Temp\\ipykernel_3416\\3480656614.py:7: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 992, in launch_instance\n",
      "    app.start()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 736, in start\n",
      "    self.io_loop.start()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 195, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 516, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 505, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 412, in dispatch_shell\n",
      "    await result\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 740, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 422, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 546, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3024, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3079, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3284, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3466, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3526, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\hp\\AppData\\Local\\Temp\\ipykernel_3416\\3861233590.py\", line 42, in <module>\n",
      "    client = MongoClient(uri, server_api=ServerApi('1'))\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  client = MongoClient(uri, server_api=ServerApi('1'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bad auth : authentication failed, full error: {'ok': 0, 'errmsg': 'bad auth : authentication failed', 'code': 8000, 'codeName': 'AtlasError'}\n"
     ]
    }
   ],
   "source": [
    "from pymongo.mongo_client import MongoClient\n",
    "from pymongo.server_api import ServerApi\n",
    "\n",
    "uri = \"mongodb+srv://algodemon106:<NetworkNinjas>@cluster0.ewjul.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0\"\n",
    "\n",
    "# Create a new client and connect to the server\n",
    "client = MongoClient(uri, server_api=ServerApi('1'))\n",
    "\n",
    "# Send a ping to confirm a successful connection\n",
    "try:\n",
    "    client.admin.command('ping')\n",
    "    print(\"Pinged your deployment. You successfully connected to MongoDB!\")\n",
    "except Exception as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "MongoClient.close() missing 1 required positional argument: 'self'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[106], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m MongoClient\u001b[38;5;241m.\u001b[39mclose()\n",
      "\u001b[1;31mTypeError\u001b[0m: MongoClient.close() missing 1 required positional argument: 'self'"
     ]
    }
   ],
   "source": [
    "MongoClient.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\hp\\AppData\\Local\\Temp\\ipykernel_3416\\350911916.py:44: ResourceWarning: Unclosed MongoClient opened at:\n",
      "  File \"<frozen runpy>\", line 198, in _run_module_as_main\n",
      "  File \"<frozen runpy>\", line 88, in _run_code\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\traitlets\\config\\application.py\", line 992, in launch_instance\n",
      "    app.start()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelapp.py\", line 736, in start\n",
      "    self.io_loop.start()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\tornado\\platform\\asyncio.py\", line 195, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\asyncio\\base_events.py\", line 607, in run_forever\n",
      "    self._run_once()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\asyncio\\base_events.py\", line 1922, in _run_once\n",
      "    handle._run()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\asyncio\\events.py\", line 80, in _run\n",
      "    self._context.run(self._callback, *self._args)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 516, in dispatch_queue\n",
      "    await self.process_one()\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 505, in process_one\n",
      "    await dispatch(*args)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 412, in dispatch_shell\n",
      "    await result\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\kernelbase.py\", line 740, in execute_request\n",
      "    reply_content = await reply_content\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\ipkernel.py\", line 422, in do_execute\n",
      "    res = shell.run_cell(\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\ipykernel\\zmqshell.py\", line 546, in run_cell\n",
      "    return super().run_cell(*args, **kwargs)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3024, in run_cell\n",
      "    result = self._run_cell(\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3079, in _run_cell\n",
      "    result = runner(coro)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\async_helpers.py\", line 129, in _pseudo_sync_runner\n",
      "    coro.send(None)\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3284, in run_cell_async\n",
      "    has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3466, in run_ast_nodes\n",
      "    if await self.run_code(code, result, async_=asy):\n",
      "  File \"c:\\Users\\hp\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py\", line 3526, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"C:\\Users\\hp\\AppData\\Local\\Temp\\ipykernel_3416\\2780173006.py\", line 44, in <module>\n",
      "    client = MongoClient(uri, server_api=ServerApi('1'))\n",
      "Call MongoClient.close() to safely shut down your client and free up resources.\n",
      "  client = MongoClient(uri, server_api=ServerApi('1'))\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "An error occurred: bad auth : authentication failed, full error: {'ok': 0, 'errmsg': 'bad auth : authentication failed', 'code': 8000, 'codeName': 'AtlasError'}\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "MongoClient.close() missing 1 required positional argument: 'self'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[107], line 61\u001b[0m\n\u001b[0;32m     58\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[0;32m     59\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAn error occurred:\u001b[39m\u001b[38;5;124m\"\u001b[39m, e)\n\u001b[1;32m---> 61\u001b[0m MongoClient\u001b[38;5;241m.\u001b[39mclose()\n",
      "\u001b[1;31mTypeError\u001b[0m: MongoClient.close() missing 1 required positional argument: 'self'"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "from pymongo import MongoClient\n",
    "from pymongo.mongo_client import MongoClient\n",
    "from pymongo.server_api import ServerApi\n",
    "\n",
    "# MongoDB connection string\n",
    "uri = \"mongodb+srv://algodemon106:NetworkNinjas@cluster0.ewjul.mongodb.net/testDatabase?retryWrites=true&w=majority\"\n",
    "\n",
    "# Function to convert MongoDB data to CSV\n",
    "def convert_to_csv(data, file_name):\n",
    "    with open(file_name, mode='w', newline='', encoding='utf-8') as file:\n",
    "        writer = csv.writer(file)\n",
    "        \n",
    "        # Write the headers\n",
    "        headers = [\n",
    "            '_id', 'userId', 'cartId', 'cartItems', 'addressInfo', 'orderStatus',\n",
    "            'paymentMethod', 'paymentStatus', 'totalAmount', 'orderDate',\n",
    "            'orderUpdateDate', 'paymentId', 'payerId', '__v'\n",
    "        ]\n",
    "        writer.writerow(headers)\n",
    "\n",
    "        # Write each document to the CSV file\n",
    "        for document in data:\n",
    "            row = [\n",
    "                str(document.get('_id', '')),\n",
    "                document.get('userId', ''),\n",
    "                document.get('cartId', ''),\n",
    "                str(document.get('cartItems', '')),\n",
    "                str(document.get('addressInfo', '')),\n",
    "                document.get('orderStatus', ''),\n",
    "                document.get('paymentMethod', ''),\n",
    "                document.get('paymentStatus', ''),\n",
    "                document.get('totalAmount', {}).get('$numberDouble', ''),\n",
    "                document.get('orderDate', {}).get('$date', {}).get('$numberLong', ''),\n",
    "                document.get('orderUpdateDate', {}).get('$date', {}).get('$numberLong', ''),\n",
    "                document.get('paymentId', ''),\n",
    "                document.get('payerId', ''),\n",
    "                document.get('__v', {}).get('$numberInt', '')\n",
    "            ]\n",
    "            writer.writerow(row)\n",
    "\n",
    "# Connect to MongoDB and fetch data\n",
    "file_name = \"output.csv\"\n",
    "client = MongoClient(uri, server_api=ServerApi('1'))\n",
    "try:\n",
    "    # Test connection\n",
    "    client.admin.command('ping')\n",
    "    print(\"Successfully connected to MongoDB!\")\n",
    "\n",
    "    # Access database and collection\n",
    "    db = client['testDatabase']\n",
    "    collection = db['orders']  # Replace 'orders' with your collection name\n",
    "    data = collection.find()\n",
    "\n",
    "    # Export data to CSV\n",
    "    convert_to_csv(data, file_name)\n",
    "    print(f\"Data successfully exported to {file_name}\")\n",
    "except Exception as e:\n",
    "    print(\"An error occurred:\", e)\n",
    "\n",
    "MongoClient.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import numpy as np\n",
    "\n",
    "# Load dataset (replace 'file_path' with your actual file path, )\n",
    "data = pd.read_csv('datasets/Amazon Sale Report.csv', low_memory=False)\n",
    "\n",
    "# Preprocessing\n",
    "# Parse the 'Date' column as datetime\n",
    "data['Date'] = pd.to_datetime(data['Date'], errors='coerce')\n",
    "\n",
    "# Remove rows with missing 'Date' or 'Qty'\n",
    "data = data.dropna(subset=['Date', 'Qty'])\n",
    "\n",
    "# Aggregate data by month and category\n",
    "monthly_data = data.groupby([data['Date'].dt.to_period('M'), 'Category']).agg(\n",
    "    total_qty=('Qty', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "# Convert 'Date' period to timestamp\n",
    "monthly_data['Date'] = monthly_data['Date'].dt.to_timestamp()\n",
    "\n",
    "# One-hot encode the 'Category' column\n",
    "data_encoded = pd.get_dummies(monthly_data, columns=['Category'], drop_first=True)\n",
    "\n",
    "# Features and target\n",
    "X = data_encoded.drop(columns=['total_qty', 'Date'])\n",
    "y = data_encoded['total_qty']\n",
    "\n",
    "# Train-test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Train a Random Forest Regressor\n",
    "model = RandomForestRegressor(random_state=42, n_estimators=100)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "# Predictions and evaluation\n",
    "y_pred = model.predict(X_test)\n",
    "rmse = np.sqrt(mean_squared_error(y_test, y_pred))\n",
    "print(f\"Model RMSE: {rmse}\")\n",
    "\n",
    "# Forecasting for the next month\n",
    "# Prepare input for the next month's prediction\n",
    "latest_data = monthly_data[monthly_data['Date'] == monthly_data['Date'].max()]\n",
    "latest_encoded = pd.get_dummies(latest_data, columns=['Category'], drop_first=True).drop(columns=['total_qty', 'Date'])\n",
    "\n",
    "# Ensure alignment with training data columns\n",
    "missing_cols = set(X.columns) - set(latest_encoded.columns)\n",
    "for col in missing_cols:\n",
    "    latest_encoded[col] = 0\n",
    "\n",
    "latest_encoded = latest_encoded[X.columns]  # Reorder columns\n",
    "\n",
    "# Predict the next month's sales quantity\n",
    "next_month_prediction = model.predict(latest_encoded)\n",
    "\n",
    "# Output predictions\n",
    "predictions = latest_data[['Category']].copy()\n",
    "predictions['Predicted_Qty'] = next_month_prediction\n",
    "\n",
    "print(\"Predicted sales quantity for the next month:\")\n",
    "print(predictions)#['Predicted_Qty'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model RMSE: 182.70664107646564\n",
      "Predicted sales quantity for the next month:\n",
      "               Category  Predicted_Quantity\n",
      "5    Festive Essentials           42.050333\n",
      "6  Traditional Clothing           17.078667\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import numpy as np\n",
    "\n",
    "# Load dataset\n",
    "data = pd.read_csv('datasets/Indian_Artisan_Data.csv', low_memory=False)\n",
    "\n",
    "# Preprocessing\n",
    "# Parse the 'Date' column as datetime\n",
    "data['Date'] = pd.to_datetime(data['Date'], errors='coerce')\n",
    "\n",
    "# Remove rows with missing 'Date' or 'Quantity'\n",
    "data = data.dropna(subset=['Date', 'Quantity'])\n",
    "\n",
    "# Aggregate data by month and category\n",
    "monthly_data = data.groupby([data['Date'].dt.to_period('M'), 'Category']).agg(\n",
    "    total_quantity=('Quantity', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "# Convert 'Date' period to timestamp\n",
    "monthly_data['Date'] = monthly_data['Date'].dt.to_timestamp()\n",
    "\n",
    "# One-hot encode the 'Category' column\n",
    "data_encoded = pd.get_dummies(monthly_data, columns=['Category'], drop_first=True)\n",
    "\n",
    "# Features and target\n",
    "X = data_encoded.drop(columns=['total_quantity', 'Date'])\n",
    "y = data_encoded['total_quantity']\n",
    "\n",
    "# Train-test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Train a Random Forest Regressor\n",
    "model = RandomForestRegressor(random_state=42, n_estimators=100)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "# Predictions and evaluation\n",
    "y_pred = model.predict(X_test)\n",
    "rmse = np.sqrt(mean_squared_error(y_test, y_pred))\n",
    "print(f\"Model RMSE: {rmse}\")\n",
    "\n",
    "# Forecasting for the next month\n",
    "# Prepare input for the next month's prediction\n",
    "latest_data = monthly_data[monthly_data['Date'] == monthly_data['Date'].max()]\n",
    "latest_encoded = pd.get_dummies(latest_data, columns=['Category'], drop_first=True).drop(columns=['total_quantity', 'Date'])\n",
    "\n",
    "# Ensure alignment with training data columns\n",
    "missing_cols = set(X.columns) - set(latest_encoded.columns)\n",
    "for col in missing_cols:\n",
    "    latest_encoded[col] = 0\n",
    "\n",
    "latest_encoded = latest_encoded[X.columns]  # Reorder columns\n",
    "\n",
    "# Predict the next month's sales quantity\n",
    "next_month_prediction = model.predict(latest_encoded)\n",
    "\n",
    "# Output predictions\n",
    "predictions = latest_data[['Category']].copy()\n",
    "predictions['Predicted_Quantity'] = next_month_prediction\n",
    "\n",
    "print(\"Predicted sales quantity for the next month:\")\n",
    "print(predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model RMSE: 44.835784536461496\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'Product Name'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32mc:\\Users\\hp\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[1;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\\\_libs\\\\hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\\\_libs\\\\hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Product Name'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[117], line 61\u001b[0m\n\u001b[0;32m     59\u001b[0m \u001b[38;5;66;03m# Create a dictionary to store predictions\u001b[39;00m\n\u001b[0;32m     60\u001b[0m predictions_dict \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m---> 61\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m product_name, predicted_qty \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mzip\u001b[39m(latest_data[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mProduct Name\u001b[39m\u001b[38;5;124m'\u001b[39m], next_month_prediction):\n\u001b[0;32m     62\u001b[0m     predictions_dict[product_name] \u001b[38;5;241m=\u001b[39m predicted_qty\n\u001b[0;32m     63\u001b[0m \u001b[38;5;66;03m# Output predictions\u001b[39;00m\n",
      "File \u001b[1;32mc:\\Users\\hp\\anaconda3\\Lib\\site-packages\\pandas\\core\\frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mget_loc(key)\n\u001b[0;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[1;32mc:\\Users\\hp\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[0;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[0;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[0;32m   3810\u001b[0m     ):\n\u001b[0;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[1;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[0;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[0;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'Product Name'"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import numpy as np\n",
    "\n",
    "# Load dataset\n",
    "data = pd.read_csv('datasets/Indian_Artisan_Data.csv', low_memory=False)\n",
    "\n",
    "# Preprocessing\n",
    "# Parse the 'Date' column as datetime\n",
    "data['Date'] = pd.to_datetime(data['Date'], errors='coerce')\n",
    "\n",
    "# Remove rows with missing 'Date' or 'Quantity'\n",
    "data = data.dropna(subset=['Date', 'Quantity'])\n",
    "\n",
    "# Aggregate data by month, category, and sub-category\n",
    "monthly_data = data.groupby([data['Date'].dt.to_period('M'), 'Category', 'Sub Category']).agg(\n",
    "    total_quantity=('Quantity', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "# Convert 'Date' period to timestamp\n",
    "monthly_data['Date'] = monthly_data['Date'].dt.to_timestamp()\n",
    "\n",
    "# One-hot encode the 'Category' and 'Sub Category' columns\n",
    "data_encoded = pd.get_dummies(monthly_data, columns=['Category', 'Sub Category'], drop_first=True)\n",
    "\n",
    "# Features and target\n",
    "X = data_encoded.drop(columns=['total_quantity', 'Date'])\n",
    "y = data_encoded['total_quantity']\n",
    "\n",
    "# Train-test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Train a Random Forest Regressor\n",
    "model = RandomForestRegressor(random_state=42, n_estimators=100)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "# Predictions and evaluation\n",
    "y_pred = model.predict(X_test)\n",
    "rmse = np.sqrt(mean_squared_error(y_test, y_pred))\n",
    "print(f\"Model RMSE: {rmse}\")\n",
    "\n",
    "# Forecasting for the next month\n",
    "# Prepare input for the next month's prediction\n",
    "latest_data = monthly_data[monthly_data['Date'] == monthly_data['Date'].max()]\n",
    "latest_encoded = pd.get_dummies(latest_data, columns=['Category', 'Sub Category'], drop_first=True).drop(columns=['total_quantity', 'Date'])\n",
    "\n",
    "# Ensure alignment with training data columns\n",
    "missing_cols = set(X.columns) - set(latest_encoded.columns)\n",
    "for col in missing_cols:\n",
    "    latest_encoded[col] = 0\n",
    "\n",
    "latest_encoded = latest_encoded[X.columns]  # Reorder columns\n",
    "\n",
    "# Predict the next month's sales quantity\n",
    "next_month_prediction = model.predict(latest_encoded)\n",
    "\n",
    "# Create a dictionary to store predictions\n",
    "predictions_dict = {}\n",
    "for product_name, predicted_qty in zip(latest_data['Product Name'], next_month_prediction):\n",
    "    predictions_dict[product_name] = predicted_qty\n",
    "# Output predictions\n",
    "print(\"Predicted sales quantities (added to dictionary):\")\n",
    "print(predictions_dict)\n",
    "\n",
    "# Output predictions\n",
    "predictions = latest_data[['Category', 'Sub Category']].copy()\n",
    "predictions['Predicted_Quantity'] = next_month_prediction\n",
    "\n",
    "print(\"Predicted sales quantity for the next month (by Sub-Category):\")\n",
    "print(predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model RMSE: 53.27935716579171\n",
      "Predicted sales quantity for the next month (by Product Name):\n",
      "   Product Name  Predicted_Quantity\n",
      "15        Rakhi                9.57\n",
      "16        Saree                4.04\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import numpy as np\n",
    "\n",
    "# Load dataset\n",
    "data = pd.read_csv('datasets/Indian_Artisan_Data.csv', low_memory=False)\n",
    "\n",
    "# Preprocessing\n",
    "# Parse the 'Date' column as datetime\n",
    "data['Date'] = pd.to_datetime(data['Date'], errors='coerce')\n",
    "\n",
    "# Remove rows with missing 'Date' or 'Quantity'\n",
    "data = data.dropna(subset=['Date', 'Quantity'])\n",
    "\n",
    "# Aggregate data by month and Product Name\n",
    "monthly_data = data.groupby([data['Date'].dt.to_period('M'), 'Product Name']).agg(\n",
    "    total_quantity=('Quantity', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "# Convert 'Date' period to timestamp\n",
    "monthly_data['Date'] = monthly_data['Date'].dt.to_timestamp()\n",
    "\n",
    "# One-hot encode the 'Product Name' column\n",
    "data_encoded = pd.get_dummies(monthly_data, columns=['Product Name'], drop_first=True)\n",
    "\n",
    "# Features and target\n",
    "X = data_encoded.drop(columns=['total_quantity', 'Date'])\n",
    "y = data_encoded['total_quantity']\n",
    "\n",
    "# Train-test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Train a Random Forest Regressor\n",
    "model = RandomForestRegressor(random_state=42, n_estimators=100)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "# Predictions and evaluation\n",
    "y_pred = model.predict(X_test)\n",
    "rmse = np.sqrt(mean_squared_error(y_test, y_pred))\n",
    "print(f\"Model RMSE: {rmse}\")\n",
    "\n",
    "# Forecasting for the next month\n",
    "# Prepare input for the next month's prediction\n",
    "latest_data = monthly_data[monthly_data['Date'] == monthly_data['Date'].max()]\n",
    "latest_encoded = pd.get_dummies(latest_data, columns=['Product Name'], drop_first=True).drop(columns=['total_quantity', 'Date'])\n",
    "\n",
    "# Ensure alignment with training data columns\n",
    "missing_cols = set(X.columns) - set(latest_encoded.columns)\n",
    "for col in missing_cols:\n",
    "    latest_encoded[col] = 0\n",
    "\n",
    "latest_encoded = latest_encoded[X.columns]  # Reorder columns\n",
    "\n",
    "# Predict the next month's sales quantity\n",
    "next_month_prediction = model.predict(latest_encoded)\n",
    "\n",
    "# Output predictions\n",
    "predictions = latest_data[['Product Name']].copy()\n",
    "predictions['Predicted_Quantity'] = next_month_prediction\n",
    "\n",
    "print(\"Predicted sales quantity for the next month (by Product Name):\")\n",
    "print(predictions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model RMSE: 53.27935716579171\n",
      "Predicted sales quantities (added to dictionary):\n",
      "{'Rakhi': 9.57, 'Saree': 4.04}\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from sklearn.metrics import mean_squared_error\n",
    "import numpy as np\n",
    "\n",
    "# Load dataset\n",
    "data = pd.read_csv('datasets/Indian_Artisan_Data.csv', low_memory=False)\n",
    "\n",
    "# Preprocessing\n",
    "# Parse the 'Date' column as datetime\n",
    "data['Date'] = pd.to_datetime(data['Date'], errors='coerce')\n",
    "\n",
    "# Remove rows with missing 'Date' or 'Quantity'\n",
    "data = data.dropna(subset=['Date', 'Quantity'])\n",
    "\n",
    "# Aggregate data by month and Product Name\n",
    "monthly_data = data.groupby([data['Date'].dt.to_period('M'), 'Product Name']).agg(\n",
    "    total_quantity=('Quantity', 'sum')\n",
    ").reset_index()\n",
    "\n",
    "# Convert 'Date' period to timestamp\n",
    "monthly_data['Date'] = monthly_data['Date'].dt.to_timestamp()\n",
    "\n",
    "# One-hot encode the 'Product Name' column\n",
    "data_encoded = pd.get_dummies(monthly_data, columns=['Product Name'], drop_first=True)\n",
    "\n",
    "# Features and target\n",
    "X = data_encoded.drop(columns=['total_quantity', 'Date'])\n",
    "y = data_encoded['total_quantity']\n",
    "\n",
    "# Train-test split\n",
    "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)\n",
    "\n",
    "# Train a Random Forest Regressor\n",
    "model = RandomForestRegressor(random_state=42, n_estimators=100)\n",
    "model.fit(X_train, y_train)\n",
    "\n",
    "# Predictions and evaluation\n",
    "y_pred = model.predict(X_test)\n",
    "rmse = np.sqrt(mean_squared_error(y_test, y_pred))\n",
    "print(f\"Model RMSE: {rmse}\")\n",
    "\n",
    "# Forecasting for the next month\n",
    "# Prepare input for the next month's prediction\n",
    "latest_data = monthly_data[monthly_data['Date'] == monthly_data['Date'].max()]\n",
    "latest_encoded = pd.get_dummies(latest_data, columns=['Product Name'], drop_first=True).drop(columns=['total_quantity', 'Date'])\n",
    "\n",
    "# Ensure alignment with training data columns\n",
    "missing_cols = set(X.columns) - set(latest_encoded.columns)\n",
    "for col in missing_cols:\n",
    "    latest_encoded[col] = 0\n",
    "\n",
    "latest_encoded = latest_encoded[X.columns]  # Reorder columns\n",
    "\n",
    "# Predict the next month's sales quantity\n",
    "next_month_prediction = model.predict(latest_encoded)\n",
    "\n",
    "# Create a dictionary to store predictions\n",
    "predictions_dict = {}\n",
    "for product_name, predicted_qty in zip(latest_data['Product Name'], next_month_prediction):\n",
    "    predictions_dict[product_name] = predicted_qty\n",
    "\n",
    "# Output predictions\n",
    "print(\"Predicted sales quantities (added to dictionary):\")\n",
    "print(predictions_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
